subscription/index.ts

```diff
@@ -0,0 +1,100 @@
+import { useRef } from 'react'
+import useSWR from 'swr'
+import {
+  withMiddleware,
+  Key,
+  SWRHook,
+  Middleware,
+  serialize,
+  SWRConfiguration,
+  useIsomorphicLayoutEffect
+} from 'swr/_internal'
+
+export type SWRSubscription<Data = any, Error = any> = (
+  key: Key,
+  { next }: { next: (err?: Error, data?: Data) => void }
+) => void
+
+export type SWRSubscriptionResponse<Data = any, Error = any> = {
+  data?: Data
+  error?: Error
+}
+
+export type SWRSubscriptionHook<Data = any, Error = any> = (
+  key: Key,
+  subscribe: SWRSubscription<Data, Error>,
+  config?: SWRConfiguration
+) => SWRSubscriptionResponse<Data, Error>
+
+const subscriptions = new Map<Key, number>()
+const disposers = new Map()
+
+export const subscription = (<Data, Error>(useSWRNext: SWRHook) =>
+  (
+    _key: Key,
+    subscribe: SWRSubscription<Data, Error>,
+    config?: SWRConfiguration
+  ): SWRSubscriptionResponse<Data, Error> => {
+    const [key] = serialize(_key)
+    const swr = useSWRNext(key, null, config)
+    const subscribeRef = useRef(subscribe)
+
+    useIsomorphicLayoutEffect(() => {
+      subscribeRef.current = subscribe
+    })
+
+    useIsomorphicLayoutEffect(() => {
+      subscriptions.set(key, (subscriptions.get(key) || 0) + 1)
```

Comment:
```suggestion
      if (!key) return
      subscriptions.set(key, (subscriptions.get(key) || 0) + 1)
```

Necessary to return early if key is falsy?

Reply:
@shuding but will this line still create an unnecessary empty string key in cache?
https://github.com/vercel/swr/pull/1263/files#diff-87add53ebf9f9528edf5d95a0de40ea324d6d985abf5532ffb03363d58fa922cR39